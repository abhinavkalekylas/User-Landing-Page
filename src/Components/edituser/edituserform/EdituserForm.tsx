import React, { useState, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import { Form, FormGroup, Button } from "reactstrap";
import axios from "axios";
import "./EdituserForm.scss";
import {
  User,
  GenderType,
  StatusType,
  validateEmail,
  token,
} from "../../../modalfunction/Modal";

const EdituserForm = ({
  userData,
  closeModal,
}: {
  userData: User;
  closeModal: () => void;
}) => {
  const navigate = useNavigate();

  // form for user data after click edit user button
  const [userForm, setUserForm] = useState(userData);
  const [error, setError] = useState(false);

  useEffect(() => {
    setUserForm(userData);
  }, [userData]);

  const handleChange = (e: { target: { name: string; value: string } }) => {
    setUserForm({
      ...userForm,
      [e.target.name]: e.target.value,
    });
  };

  const handleForm = async (e: { preventDefault: () => void }) => {
    e.preventDefault();
    if (userForm.name === "" || userForm.email === "") {
      return alert("Please enter required field");
    } else if (!validateEmail(userForm.email)) {
      return alert("Email is not valid");
    } else {
      try {
        await axios(`https://gorest.co.in/public/v2/users/${userForm.id}`, {
          method: "PUT",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
          data: JSON.stringify(userForm),
        });
        closeModal();
        navigate(0);
      } catch (error) {
        setError(true);
        console.log("Error while editing user ", error);
      }
    }
  };
  return (
    <div className="edituser_container">
      {error ? (
        <h2>This page is under develop. We will sure give you an update</h2>
      ) : (
        <>
          <Form onSubmit={handleForm} className="form">
            <FormGroup>
              <div className="input input-group input-group-lg">
                <label className="edituser_label">Name :</label>
                <input
                  className="username form-control my-2"
                  placeholder="Enter username here"
                  aria-label="Sizing example input"
                  aria-describedby="inputGroup-sizing-lg"
                  onChange={handleChange}
                  type="text"
                  name="name"
                  value={userForm.name}
                  autoComplete="off"
                  required
                />
              </div>
              <div className="input input-group input-group-lg">
                <label className="edituser_label">Email :</label>
                <input
                  className="email form-control my-2"
                  placeholder="Enter email id here"
                  aria-label="Sizing example input"
                  aria-describedby="inputGroup-sizing-lg"
                  onChange={handleChange}
                  type="text"
                  name="email"
                  value={userForm.email}
                  autoComplete="off"
                  required
                />
              </div>
              <div className="input input-group input-group-lg">
                <label className="edituser_label">Gender :</label>
                <select
                  name="gender"
                  value={userForm.gender}
                  onChange={handleChange}
                  className="gender form-select form-select-lg mb-3"
                  aria-label=".form-select-lg example"
                >
                  <option value={GenderType.MALE}>Male</option>
                  <option value={GenderType.FEMALE}>Female</option>
                </select>
              </div>

              <div className="input input-group input-group-lg">
                <label className="edituser_label">Status :</label>
                <select
                  name="status"
                  value={userForm.status}
                  onChange={handleChange}
                  className="status form-select form-select-lg mb-3"
                  aria-label=".form-select-lg example"
                >
                  <option value={StatusType.ACTIVE}>Active</option>
                  <option value={StatusType.INACTIVE}>Inactive</option>
                </select>
              </div>
              <div className="d-flex">
                <Button
                  className="edituser_btn btn_edit"
                  color="success"
                  type="submit"
                >
                  Edit User
                </Button>
              </div>
            </FormGroup>
          </Form>
        </>
      )}
    </div>
  );
};

export default EdituserForm;
