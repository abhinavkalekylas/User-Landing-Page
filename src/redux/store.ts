import { legacy_createStore, compose } from "redux";
import reducers from "./reducer/todo"

// declare global {
//     interface Window {
//       __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
//     }
//   }

// const store = legacy_createStore(reducers, window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose);

const store = legacy_createStore(reducers, {});

export default store;