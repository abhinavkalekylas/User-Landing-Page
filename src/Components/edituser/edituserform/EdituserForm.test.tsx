import axios from "axios";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import {
  GenderType,
  StatusType,
  token,
  User,
} from "../../../modalfunction/Modal";
import store from "../../../redux/store";
import EdituserForm from "./EdituserForm";

var MockAdapter = require("axios-mock-adapter");
const userData: User = {
  id: 0,
  name: "shivam sharma",
  email: "shivam@gmail.com",
  gender: GenderType.MALE,
  status: StatusType.ACTIVE,
};

let wrapper: any;
const props = {
  userData,
  closeModal: jest.fn(),
};
beforeEach(() => {
  wrapper = mount(
    <Provider store={store}>
    <Router>
      <EdituserForm {...props} />
    </Router>
    </Provider>
  );
});

afterEach(() => {
  jest.clearAllMocks();
});

var mock = new MockAdapter(axios);

const user_data_before: User = {
  id: 123,
  name: "Harinarayan Abbott",
  email: "abbott_harinarayan@collins.info",
  gender: GenderType.MALE,
  status: StatusType.ACTIVE,
};
const user_data_after: User = {
  id: 123,
  name: "rinarayan Abbott",
  email: "bott_harinarayan@collins.info",
  gender: GenderType.MALE,
  status: StatusType.ACTIVE,
};

// snapshot testing
it("renders correctly EdituserForm component", () => {
  expect(toJson(wrapper)).toMatchSnapshot();
});

it("test username input", () => {
  const event = { target: { name: "name", value: "shivam" } };
  wrapper.find(".username").simulate("change", event);
  wrapper.update();
  expect(wrapper.find(".username").props().value).toEqual("shivam");
});

// test input tag for email
it("test email input", () => {
  const event = { target: { name: "email", value: "shivam@gmail.com" } };
  wrapper.find(".email").simulate("change", event);
  expect(wrapper.find(".email").props().value).toEqual("shivam@gmail.com");
});

// test select tag for gender
it("test gender select input", () => {
  wrapper.find(".gender").simulate("change", {
    target: { name: "gender", value: GenderType.FEMALE },
  });
  expect(wrapper.find(".gender").props().value).toBe(GenderType.FEMALE);
});

// test select tag for status
it("test status select input", () => {
  wrapper.find(".status").simulate("change", {
    target: { name: "status", value: StatusType.ACTIVE },
  });
  expect(wrapper.find(".status").props().value).toBe(StatusType.ACTIVE);
});

// Put request
describe("Put test", () => {
  it("test put request test", () => {
    const initial_user = wrapper.find("EdituserForm").props().userData;

    initial_user.status = StatusType.INACTIVE;

    wrapper.find(".status").simulate("change", {
      target: { name: "status", value: StatusType.INACTIVE },
    });

    mock
      .onPut("https://gorest.co.in/public/v2/users/")
      .reply(200, initial_user, {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      });

    expect(mock.handlers.put[0][4]).toBe(initial_user);
  });
});
