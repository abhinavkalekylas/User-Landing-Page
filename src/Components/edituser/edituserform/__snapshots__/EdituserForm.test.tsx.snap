// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly EdituserForm component 1`] = `
<BrowserRouter>
  <Router
    location={
      Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      }
    }
    navigationType="POP"
    navigator={
      Object {
        "action": "POP",
        "back": [Function],
        "block": [Function],
        "createHref": [Function],
        "forward": [Function],
        "go": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <EdituserForm
      closeModal={[MockFunction]}
      userData={
        Object {
          "email": "shivam@gmail.com",
          "gender": "male",
          "id": 0,
          "name": "shivam sharma",
          "status": "active",
        }
      }
    >
      <UserInfo
        closeModal={[MockFunction]}
        operation="edit"
        userData={
          Object {
            "email": "shivam@gmail.com",
            "gender": "male",
            "id": 0,
            "name": "shivam sharma",
            "status": "active",
          }
        }
      >
        <div
          className="userinfo_container"
        >
          <Form
            className="form"
            onSubmit={[Function]}
            tag="form"
          >
            <form
              className="form"
              onSubmit={[Function]}
            >
              <FormGroup
                tag="div"
              >
                <div
                  className="mb-3"
                >
                  <div
                    className="input input-group input-group-lg"
                  >
                    <label
                      className="userinfo_label"
                    >
                      Name :
                    </label>
                    <input
                      aria-describedby="inputGroup-sizing-lg"
                      aria-label="Sizing example input"
                      autoComplete="off"
                      className="username form-control my-2"
                      name="name"
                      onChange={[Function]}
                      placeholder="Enter username here"
                      required={true}
                      type="text"
                      value="shivam sharma"
                    />
                  </div>
                  <div
                    className="input input-group input-group-lg"
                  >
                    <label
                      className="userinfo_label"
                    >
                      Email :
                    </label>
                    <input
                      aria-describedby="inputGroup-sizing-lg"
                      aria-label="Sizing example input"
                      autoComplete="off"
                      className="email form-control my-2"
                      name="email"
                      onChange={[Function]}
                      placeholder="Enter email id here"
                      required={true}
                      type="text"
                      value="shivam@gmail.com"
                    />
                  </div>
                  <div
                    className="input input-group input-group-lg"
                  >
                    <label
                      className="userinfo_label"
                    >
                      Gender :
                    </label>
                    <select
                      aria-label=".form-select-lg example"
                      className="gender form-select form-select-lg mb-3"
                      name="gender"
                      onChange={[Function]}
                      value="male"
                    >
                      <option
                        value="male"
                      >
                        Male
                      </option>
                      <option
                        value="female"
                      >
                        Female
                      </option>
                    </select>
                  </div>
                  <div
                    className="input input-group input-group-lg"
                  >
                    <label
                      className="userinfo_label"
                    >
                      Status :
                    </label>
                    <select
                      aria-label=".form-select-lg example"
                      className="status form-select form-select-lg mb-3"
                      name="status"
                      onChange={[Function]}
                      value="active"
                    >
                      <option
                        value="active"
                      >
                        Active
                      </option>
                      <option
                        value="inactive"
                      >
                        Inactive
                      </option>
                    </select>
                  </div>
                  <div
                    className="d-flex"
                  >
                    <Button
                      className="userinfo_btn"
                      color="success"
                      tag="button"
                      type="submit"
                    >
                      <button
                        aria-label={null}
                        className="userinfo_btn btn btn-success"
                        onClick={[Function]}
                        type="submit"
                      >
                        edit
                         User
                      </button>
                    </Button>
                    <Button
                      className="reset_btn"
                      color="primary"
                      onClick={[Function]}
                      tag="button"
                    >
                      <button
                        aria-label={null}
                        className="reset_btn btn btn-primary"
                        onClick={[Function]}
                        type="button"
                      >
                        Reset
                      </button>
                    </Button>
                  </div>
                </div>
              </FormGroup>
            </form>
          </Form>
        </div>
      </UserInfo>
    </EdituserForm>
  </Router>
</BrowserRouter>
`;
